### PR TITLE
[RPM M1] Add rpm assemble/bundle code for opensearch using rpmbuild

### PR DESCRIPTION
**Pull request by peterzhuamazon**
### Description
Add rpm assemble/bundle code for opensearch using rpmbuild.
Adding rpm specific block so that cpio is used during extraction, and rpmbuild is used during build.
A specific opensearch spec file is provided, dashboards one will be committed in another PR.
 
### Issues Resolved
Part of #1545
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peterzhuamazon** [comment]
Hi @dblock please review the code and I will add test cases later.

Thanks.

File: scripts/components/OpenSearch-Dashboards/install.sh [hunk 1]
```diff
@@ -66,11 +70,22 @@ fi
...
     DIR="$(dirname "$0")"
     echo $DIR
     cd $DIR
     cp ../../../config/opensearch_dashboards.yml "$OUTPUT/config/"
 )
+elif [ "$DISTRIBUTION" = "rpm" ]; then
+(   
+    DIR="$(dirname "$0")"
+    echo $DIR
+    cd $DIR
```
Comment (dblock) [inline scripts/components/OpenSearch-Dashboards/install.sh:87]: Why is this needed? I don't expect scripts to change current directory. 

If there's a good reason, the `DIR` part is common and should be taken out of the `if`.

**Review by dblock — COMMENTED** [review]
Needs passing tests and documentation updates.

File: scripts/components/OpenSearch-Dashboards/install.sh (hunk 1 above)
Comment (peterzhuamazon) [reply scripts/components/OpenSearch-Dashboards/install.sh:87]: I forget to make it common as I was copy paste during testing, will change.

File: scripts/components/OpenSearch-Dashboards/install.sh [hunk 2]
```diff
@@ -72,20 +72,16 @@ fi
 [ -z "$ARCHITECTURE" ] && ARCHITECTURE=`uname -m`
 [ -z "$DISTRIBUTION" ] && DISTRIBUTION="tar"
 
+# Make sure the cwd is where the script is located
+DIR="$(dirname "$0")"
+echo $DIR
+cd $DIR
```
Comment (dblock) [inline scripts/components/OpenSearch-Dashboards/install.sh:78]: So, why is this necessary?

Comment (peterzhuamazon) [reply scripts/components/OpenSearch-Dashboards/install.sh:78]: If this is not there we cannot find the script folder location.
If you call install script outside of its located folder, it will default cwd to other folder.

File: scripts/components/OpenSearch-Dashboards/install.sh (hunk 1 above)
Comment (peterzhuamazon) [reply scripts/components/OpenSearch-Dashboards/install.sh]: fixed.

File: scripts/components/OpenSearch-Dashboards/install.sh (hunk 2 above)
Comment (dblock) [reply scripts/components/OpenSearch-Dashboards/install.sh:78]: Changing directories is generally not a good pattern, because it relies on that location. You could use `$DIR/script.sh` to call it for example. The script should probably be agnostic of where it's called from. 

It's not a huge deal, but generally having to `cd` hides bugs in my experience.

Comment (peterzhuamazon) [reply scripts/components/OpenSearch-Dashboards/install.sh:78]: This is due to the fact we are using relative paths right now for the copy.
A lot of the scripts we have right now is relying on cd.
Will definitely worth revisit this later.

Thanks.

Comment (dblock) [reply scripts/components/OpenSearch-Dashboards/install.sh:78]: Where? Which scripts? I cannot find any instances of such.

Even if that's the case, can we please call these scripts from the right directory as a workaround instead of "fixing" it this way? I am worried this propagates into dozens of scripts and becomes the default pattern that we can never undo. I am 100% sure no other build.sh/install.sh does this and we're going to have to change all of them.

**Review by dblock — CHANGES_REQUESTED** [review]
The `cd` worries me. Can we merge without that?

File: scripts/components/OpenSearch-Dashboards/install.sh (hunk 2 above)
Comment (peterzhuamazon) [reply scripts/components/OpenSearch-Dashboards/install.sh:78]: @dblock I did not add these lines, these are existing lines and we have been using this method since ODFE.

Examples:
https://github.com/opensearch-project/opensearch-build/blob/main/scripts/opensearch-onetime-setup.sh#L12
https://github.com/opensearch-project/opensearch-build/blob/main/scripts/deploy_single_tar_cluster.sh#L18
https://github.com/opensearch-project/opensearch-build/blob/main/scripts/components/OpenSearch-Dashboards/install.sh#L72-L74

**Comment by peterzhuamazon** [comment]
@spotrh would you mind checking the spec file in this PR?

Thanks.

Note: as of now we are using bundled JDK per https://github.com/opensearch-project/opensearch-build/issues/1682#issuecomment-1059525252

And we are also thinking about using Adopt per community suggestions: https://github.com/opensearch-project/OpenSearch/issues/2312#issuecomment-1057524268

**Review by dblock — APPROVED** [review]

**Review by zelinh — CHANGES_REQUESTED** [review]
Block this PR for now until Spot review the spec files.

**Review by spotrh — APPROVED** [review]
Looks very good to me. I don't understand the tests well enough here to be sure, but I don't think there is any CI integration to confirm that the RPM was properly generated (or generated at all). Not a blocker for this, but probably something you want to do.

**Comment by peterzhuamazon** [comment]
> Looks very good to me. I don't understand the tests well enough here to be sure, but I don't think there is any CI integration to confirm that the RPM was properly generated (or generated at all). Not a blocker for this, but probably something you want to do.

Yes @spotrh we have plans to add tests and verifications here: #27 
See milestone 5 / 6 about the steps we plan to do.

Thanks.

**Review by zelinh — DISMISSED** [review]
Block this one from merging since we haven't generated the artifacts candidates for 1.3.0 release.
